### PR TITLE
Inline `trace` and `error` via annotations

### PR DESCRIPTION
**Pull request by zliu41**
Previously these functions are inlined by permitting `LamAbs` and `TyAbs` to be inlined. Using annotations lets us control inlining more precisely.

### CODE REVIEW

File: plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs [hunk 1]
```diff
@@ -320,12 +321,12 @@ defineBuiltinTerms = do
             ta <- freshTyName "a"
             t <- freshName "t"
             a <- freshName "a"
-            pure $ PIR.tyAbs () ta (PLC.Type ())
+            pure $ PIR.tyAbs AnnInline ta (PLC.Type AnnOther)
```
Comment (michaelpj) [inline plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs:324]: Needs some comments!

**Review by michaelpj — COMMENTED** [review]
Generally looks good, I just think we should try and keep keying the inline hints off the annotation on the binding, rather than the RHS>

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs [hunk 2]
```diff
@@ -315,7 +316,7 @@ maybeAddSubst a s n rhs = do
 
     -- Check whether we've been told specifically to inline this
     hints <- asks _hints
-    let hinted = shouldInline hints a n
+    let hinted = shouldInline hints a n (termAnn rhs')
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs:319]: Also, this is now inconsistent with the UPLC inliner!

File: plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs [hunk 3]
```diff
@@ -551,7 +551,7 @@ entryExitTracingInside ::
     PIR.Name
     -> T.Text
     -> PIRTerm PLC.DefaultUni PLC.DefaultFun
-    -> PLC.Type PIR.TyName PLC.DefaultUni ()
+    -> PLC.Type PIR.TyName PLC.DefaultUni Ann  -- (zliu41) ???
```
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:554]: Oops, I meant to revisit this to see whether this change is really needed

File: plutus-core/common/PlutusCore/InlineUtils.hs [hunk 4]
```diff
@@ -1,9 +1,18 @@
-module PlutusCore.InlineUtils (InlineHints(..)) where
+module PlutusCore.InlineUtils (InlineHints (..), ShouldInline) where
 
 import Data.Semigroup (Any (..))
 
-newtype InlineHints name a = InlineHints { shouldInline :: a -> name -> Bool }
-    deriving (Semigroup, Monoid) via (a -> name -> Any)
+type ShouldInline name a =
+    -- | Binder annotation
```
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-core/common/PlutusCore/InlineUtils.hs:6]: Let me double check, but this is how Ormolu formats haddocks so it should.

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs:319]: So there are two reasons why we look at both the binding annotation and the RHS annotation:
- Some inlining criteria (such as whether the RHS is `trace`) depends only on the RHS. So it's more convenient to just annotate the RHS with that, rather than having to propagate that annotation to the entire binding.
- The binding may already have other annotations like `DatatypeComponent`. To have additional annotations on the binding require a bigger change than this.

Does this make sense?

btw I made the same change to the UPLC inliner.

File: plutus-core/common/PlutusCore/InlineUtils.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-core/common/PlutusCore/InlineUtils.hs:6]: Yep this does work.

File: plutus-tx-plugin/src/PlutusTx/PIRTypes.hs [hunk 5]
```diff
@@ -1,13 +1,14 @@
...
 import PlutusIR qualified as PIR
+import PlutusTx.Annotation
 
-type PIRKind = PIR.Kind ()
-type PIRType uni = PIR.Type PIR.TyName uni ()
-type PIRTerm uni fun = PIR.Term PIR.TyName PIR.Name uni fun ()
+type PIRKind = PIR.Kind Ann
+type PIRType uni = PIR.Type PIR.TyName uni Ann
+type PIRTerm uni fun = PIR.Term PIR.TyName PIR.Name uni fun Ann
 type PIRProgram uni fun = PIR.Program PIR.TyName PIR.Name uni fun ()
```
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-tx-plugin/src/PlutusTx/PIRTypes.hs:9]: Currently there's no need to have annotation on PIRProgram/PLCProgram/UPLCProgram, since once have a Program, the compilation is done and the only thing we do to it is `dumpFlat`.

Should we keep it as it is until we need to annotation the Program?

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs:319]: > Some inlining criteria (such as whether the RHS is trace) depends only on the RHS

I disagree with this framing, I think. I think it really is a property of the binding: we want to inline _that binding_ not _that term_. For example:

```
let
   -- Original binding that we want to inline 
   trace = ...
   -- Just some other binding
   trace2 = trace
in trace2 ...
```

Should we inline `trace2`? I think we should not, but with your implementation we will, because the "should inline" property attaches to the term, not the binding.

> The binding may already have other annotations like DatatypeComponent. To have additional annotations on the binding would probably require a bigger change than this.

The `Provenance` annotations all have a "parent" provenance, which should eventually top out with an `Original` provenance. That is, we should be able to write a `getOriginal :: Provenance a -> a` function (okay, maybe we need to do something for `MultipleSources`, either return `Nothing` or require `Monoid a`, I guess). So we should be able to pull out the underlying annotation in all cases!

[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs:319]: @michaelpj 

> Should we inline `trace2`?

That's exactly what I'd expect, so

> I think we should not

why not?

When I just started reviewing this, I thought the same as you, however after some staring at the code I realized I like what's implemented here much more. Not only does it allow us to inline the obvious candidates like `/\a -> \text x -> trace {a} text x`, but also their instantiations like `\text (x : integer) -> trace {integer} text x` and the partial application `\(x : integer) -> trace {integer} text x` too. This means one can inline a term, then partially reduce it as its use site, inline it, compute a little more, opening another opportunity for inlining, and so forth.

It's much trickier though as we now can end up inlining closures. What if `text` is some huge chunk of text, do we then want to inline `\(x : integer) -> trace {integer} text x`? Probably not. BTW, is it ensured that we only inline `trace` when traces are ignored or is it considered fine to always inline them?

But currently we have <10 `AnnInline` mentions, so I wouldn't worry too much about the complexity of the approach. It's certainly not any more complex than any other pass that we have.

[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs:319]: The goal here is somewhat limited: we're going to make a specific binding which doesn't look like it should inline, and we want it to inline (including the type/lambda abstractions inside it). 

> But currently we have <10 AnnInline mentions, so I wouldn't worry too much about the complexity of the approach. It's certainly not any more complex than any other pass that we have.

Well, I think it's making it more complex than it was before. Previously we just looked at bindings, now we also look at terms!

[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs:319]: @michaelpj Ok, so to propagate annotations from the RHS of a binding to the binding itself, we need to do these:
- Update `runDefT`: #4637  (I think we should do this regardless, hence the PR)
- Every time we construct a `TermBind` we must remember to propagate the annotation. This can be done by a smart constructor so is not too bad, but is there really no case where we _don't_ want to propagate?
- Whenever we simplify RHS, turning `let v = RHS in expr` into `let v = RHS' in expr`, if `RHS'` has a new annotation, we must remember to modify `v`'s annotation.

The last one sounds pretty unattractive. What do you think?

File: plutus-core/common/PlutusCore/InlineUtils.hs [hunk 6]
```diff
@@ -1,9 +1,18 @@
...
+    a ->
+    -- | Binder name
+    name ->
+    -- | RHS annotation
+    a ->
+    Bool
```
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-core/common/PlutusCore/InlineUtils.hs:12]: Yep (mainly for new modules). I'm not a huge fan of Ormolu's formatting by any means, but I like manually formatting the code a lot less :smile:

File: plutus-core/plutus-ir/test/transform/inline/single.golden [hunk 7]
```diff
@@ -9,7 +9,15 @@
...
-    [ (lam y (con integer) y) [ (lam y (con integer) y) (con integer 1) ] ]
+    (let
+      (nonrec)
+      (termbind
+        (strict)
+        (vardecl f (fun (con integer) (con integer)))
+        (lam y (con integer) y)
+      )
+      [ f [ f (con integer 1) ] ]
+    )
```
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-core/plutus-ir/test/transform/inline/single.golden:20]: Not really. This PR stops allowing `LamAbs` and `TyAbs` to be inlined (except for `trace` and `error`). This could result in increased program sizes in some cases, but could also do the opposite in some other cases.

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs:319]: Okay, here's what I was thinking:
- In `Definitions`, propagate the annotation on the `VarDecl` to the `Binding` also. We're quite fuzzy about this, but I think hanging it off the variable is reasonable.
- Then we just need to make sure that the `VarDecl`s we create have the bindings that we want. That shouldn't be too bad: we can just pass an annotation into `defineBuiltinTerm` etc. and use that.

So basically I think we should keep the strategy of "indicate what annotation to use by attaching it to the data in `Definitions`" rather than "pass in functions at the end to fix up the annotations", but I think there is a good place to put the annotation.

File: plutus-benchmark/nofib/test/Spec.hs [hunk 8]
```diff
@@ -46,7 +46,7 @@ testClausify = testGroup "clausify"
                , testCase "formula3" $ mkClausifyTest Clausify.F3
                , testCase "formula4" $ mkClausifyTest Clausify.F4
                , testCase "formula5" $ mkClausifyTest Clausify.F5
-               , Tx.fitsInto "formula1 (size)" (Clausify.mkClausifyCode Clausify.F1) 4700
+               , Tx.fitsInto "formula1 (size)" (Clausify.mkClausifyCode Clausify.F1) 4720
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-benchmark/nofib/test/Spec.hs:49]: That is kind of a good question: this should only affect examples with the flag on, right?

[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-benchmark/nofib/test/Spec.hs:49]: Sorry, what am I talking about: of course things will change, because we're now less keen to inline random lambdas.

File: plutus-core/plutus-ir/src/PlutusIR/Compiler/Provenance.hs [hunk 9]
```diff
@@ -35,6 +35,15 @@ instance Ord a => Semigroup (Provenance a) where
   MultipleSources ps1 <> x                   = MultipleSources (S.insert x ps1)
   x <> y                                     = MultipleSources (S.fromList [x,y])
 
+instance Foldable Provenance where
```
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Compiler/Provenance.hs:38]: cute

**Review by michaelpj — APPROVED** [review]
